src/java/htsjdk/samtools/sra/SRAAlignmentIterator.java

```diff
@@ -0,0 +1,198 @@
+            }
+
+            refOffset += refLen;
+        }
+        referencesChunksIterator = referencesChunks.iterator();
+
+        try {
+            nextReference();
+        } catch (final Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        // check aligned
+        if (hasMoreAlignments == null) {
+            try {
+                lastRecord.detachFromIterator();
+                hasMoreAlignments = alignedIterator.nextAlignment();
+            } catch (ErrorMsg e) {
+                throw new RuntimeException(e);
+            }
+        }
+        while (!hasMoreAlignments && hasMoreReferences) {
+            nextReference();
+        }
+
+        return hasMoreAlignments;
+    }
+
+    @Override
+    public SAMRecord next() {
+        if (!hasNext()) {
+            return null;
+        }
+
+        return nextAlignment();
+    }
+
+    @Override
+    public void remove() {
+        throw new RuntimeException("Removal of records not implemented.");
+    }
+
+    public void setValidationStringency(ValidationStringency validationStringency) {
+        this.validationStringency = validationStringency;
+    }
+
+    private SAMRecord nextAlignment() {
+        try {
+            lastRecord = new SRALazyRecord(header, true, run, alignedIterator, alignedIterator.getReadId(), alignedIterator.getAlignmentId());
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+        if (validationStringency != null) {
+            lastRecord.setValidationStringency(validationStringency);
+        }
+
+        hasMoreAlignments = null;
+
+        return lastRecord;
+    }
+
+    private void nextReference() {
+        if (!hasMoreReferences) {
+            throw new RuntimeException("Cannot get next reference - already at last one");
+        }
+
+        try {
+            hasMoreReferences = referencesChunksIterator.hasNext();
+            if (!hasMoreReferences) {
+                hasMoreAlignments = false;
+                return;
+            }
+
+            currentReference++;
+            Chunk refChunk = referencesChunksIterator.next();
+            if (refChunk == null) {
+                hasMoreAlignments = false;
+                return;
+            }
+
+            Reference reference = cachedReferences.get(currentReference);
+            // just in case different iterators from the same reader are used from different threads
+            if (reference == null) {
+                synchronized (cachedReferences) {
+                    reference = cachedReferences.get(currentReference);
+                    if (reference == null) {
+                        reference = run.getReference(header.getSequence(currentReference).getSequenceName());
+                        cachedReferences.set(currentReference, reference);
+                    }
+                }
+            }
```

Comment:
This is a very questionable-looking block of code. If you want to make the caching thread-safe, you should either make the references cache `ThreadLocal`, or create a `ReferenceCache` wrapper class with synchronized methods.